### PR TITLE
User validation (Fixes #1996)

### LINKED ISSUE #1996: Add Email Validation
Hi, this is a first-timers-only issue. This means we've worked to make it more legible to folks who either haven't contributed to our codebase before, or even folks who haven't contributed to open source before.

If that's you, we're interested in helping you take the first step and can answer questions and help you out as you do. Note that we're especially interested in contributions from people from groups underrepresented in free and open source software!

We know that the process of creating a pull request is the biggest barrier for new contributors. This issue is for you :gift_heart:

If you have contributed before, consider leaving this one for someone new, and looking through our general help wanted issues. Thanks!
:thinking: What you will need to know.

Nothing. This issue is meant to welcome you to Open Source :) We are happy to walk you through the process.
## Please describe the problem

A user is able to sign in and use publiclab with email wrong email addresses say bansal@gmail

:clipboard: Step by Step

:raising_hand_woman: Claim this issue: Comment below. If someone else has claimed it, ask if they've opened a pull request already and if they're stuck -- maybe you can help them solve a problem or move it along!

:memo: Add VALID_EMAIL_REGEX = /\A[\w+-.]+@[a-z\d-.]+.[a-z]+\z/i to the 
https://github.com/publiclab/plots2/blob/3b09a3515bbfecb48194bd86f2d467b2710103be/app/models/user.rb#L16
:floppy_disk: Commit your changes

:twisted_rightwards_arrows: Start a Pull Request.

:checkered_flag: Done Ask in comments for a review :)

:thinking::question: Questions?

Leave a comment below!

We encourage you to link to this issue by mentioning the issue # in your pull request, so we can see if someone's already started on it. If someone seem stuck, offer them some help! Otherwise, take a look at some other issues you can help with. Thanks!
Thank you!

Your help makes Public Lab better! We deeply appreciate your helping refine and improve this site.

### PR DESCRIPTION
**Pull request by Ankit-Singla**
Make sure these boxes are checked before your pull request is ready to be reviewed and merged. Thanks!

* [x] all tests pass -- `rake test:all`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request is descriptively named with #number reference back to original issue

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. Please alert developers on plots-dev@googlegroups.com when your request is ready or if you need assistance.

Thanks!

### CODE REVIEW

**Comment by Ankit-Singla** [comment]
Will this do?
I forgot to rebase the feature branch on top of the master branch this time. Or should I open another pull request after doing that?
Rest everything is fine, I suppose.

**Comment by namangupta01** [comment]
@Ankit-Singla Just remove the commits of email validation and rebase it with plots master. if you need help doing so you can ask!

**Comment by Ankit-Singla** [comment]
@namangupta01 Why do I need to remove commits? And how can I do that?

**Comment by namangupta01** [comment]
@Ankit-Singla you have three commits here 0c86186, 7b41e61 and 53f3f61 where 7b41e61 and 53f3f61 belongs to your pr #2164 . We only need 0c86186 this commit here. Doesn't we?

**Comment by Ankit-Singla** [comment]
@namangupta01  
Okay, got it.
And how can I remove those commits?

**Comment by namangupta01** [comment]
For removing commits you can type` git rebase -i HEAD~3` here 3 represents the **number of previous 2 commits from the current head and including current head** and then you will get three commits info with commit hash, message and description then delete the whole hash of the commits that you want to delete. I think that will works for you.
P.S For trying and learning first and to know hows it works try the above stuff on any demo branch so that in case something went wrong your work will not be lost! :)

**Comment by Ankit-Singla** [comment]
@SidharthBansal Please review this pull request.

**Comment by ViditChitkara** [comment]
This is working perfectly here http://rubular.com/r/Pyb899szdt
Thanks @Ankit-Singla :+1:

**Comment by jywarren** [comment]
This looks great! But i wanted to note -- what about non-latin characters? Sorry because this is pretty tough, but consider:

* http://rubular.com/r/qU9il8h8SD
* http://rubular.com/r/GBFSvPBgXe

What do you think?

**Comment by Ankit-Singla** [comment]
@jywarren We can use POSIX bracket expressions to solve this issue. Sorry that I wasn't careful this time. Therefore, a better regex would be `/\A[-[:alnum:]+.]+@[\da-z-.]+[.][a-z]+\z/`, if we want to allow non-latin characters only before the `@` symbol, otherwise `/\A[-[:alnum:]+.]+@[[:alnum:]-.]+[.][[:alpha:]]+\z/` would be a better choice.
Do you approve of any of the following, so that I could make the required changes in the Pull Request?

**Comment by jywarren** [comment]
@ViditChitkara what do you think? This looks good to me!

**Comment by ViditChitkara** [comment]
Looks super!! :+1

**Comment by SidharthBansal** [comment]
Looks good to me too

**Comment by Ankit-Singla** [comment]
@jywarren So I have made the required changes to include non-latin characters in email validation. Please review the pull request.

**Comment by jywarren** [comment]
This is an odd error in Travis:

```
1.72s$ if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then danger --verbose; fi
You used `puts` in your Dangerfile. To print out text to GitHub use `message` instead
There was an error with Danger bot's Junit parsing:
undefined method `value' for nil:NilClass
⚠️ 1 Warning. Don't worry, everything is fixable.
Danger does not have write access to the PR to set a PR status.
```

https://travis-ci.org/publiclab/plots2/builds/337619181#L3377

I think we have a problem in our Dangerfile that's preventing it from reporting out the error correctly.

File: test/unit/user_test.rb [hunk 1]
```diff
@@ -146,6 +146,13 @@ class UserTest < ActiveSupport::TestCase
     assert_equal 1, user.errors[:email].count
   end
 
+  test 'user email validation' do
+    user = User.new(username: 'ankit',
+                    password: 'test1234',
+                    password_confirmation: 'test1234',
+                    email: 'ankit@gmail')
+    assert_not user.save({})
```
Comment (jywarren) [inline test/unit/user_test.rb]: Ah! Oops, you forgot an `end` here -- that should fix it!

Comment (jywarren) [reply test/unit/user_test.rb]: @Ankit-Singla everything all right? I think adding `end` will fix this and we'll be good to go. Thank you!

Comment (jywarren) [reply test/unit/user_test.rb]: Hi @Ankit-Singla -- no rush, just making sure you've seen this and if you need help we can pitch in! :-)

**Comment by SidharthBansal** [comment]
Issue fixed #2506

**Pull request closed without merging**